toolchain/semantics/semantics_ir.cpp

```diff
@@ -286,6 +288,12 @@ auto SemanticsIR::StringifyType(SemanticsTypeId type_id,
 
     auto node = GetNode(step.node_id);
     switch (node.kind()) {
+      case SemanticsNodeKind::ArrayType: {
+        auto [bound_id, type_id] = node.GetAsArrayType();
+        out << "array [" << StringifyType(type_id) << "; "
+            << GetArrayBound(bound_id) << "]";
+        break;
+      }
```

Comment:
We want this function to be non-recursive, so that very complex types can't overflow our stack. Look at the implementation for `ConstType` to see how to print an inner type (the element type in this case) with some extra text wrapped around it (the `[` and `; bound]`).